spark/src/main/java/org/apache/iceberg/spark/SparkUtil.java

```diff
@@ -99,4 +103,30 @@ public static void validatePartitionTransforms(PartitionSpec spec) {
       }
     }
   }
+
+  /**
+   * Pulls any Catalog specific overrides for the Hadoop conf from the current SparkSession, which can be
+   * set via spark.sql.catalog.$catalogName.hadoop.*
+   *
+   * The SparkCatalog allows for hadoop configurations to be overridden per catalog, by setting
+   * them on the SQLConf, where the following will add the property "fs.default.name" with value
+   * "hdfs://hanksnamenode:8020" to the catalog's hadoop configuration.
+   *   SparkSession.builder()
+   *     .config(s"spark.sql.catalog.$catalogName.hadoop.fs.default.name", "hdfs://hanksnamenode:8020")
+   *     .getOrCreate()
+   * @param spark The current Spark session
+   * @param catalogName Name of the catalog to find overrides for.
+   * @return the Hadoop Configuration that should be used for this catalog, with catalog specific overrides applied.
+   */
+  public static Configuration hadoopConfCatalogOverrides(SparkSession spark, String catalogName) {
+    // Find keys for the catalog intended to be hadoop configurations
+    final String hadoopConfCatalogPrefix = String.format("%s.%s.%s", ICEBERG_CATALOG_PREFIX, catalogName, "hadoop.");
+    Configuration conf = spark.sessionState().newHadoopConf();
+    spark.sqlContext().conf().settings().forEach((k, v) -> {
+      if (v != null && k.startsWith(hadoopConfCatalogPrefix)) {
```

Comment:
Do we need to check if v is null? It's OK to pass the null, right?

Reply:
We need to check if k is null, otherwise it hits a NPE.

Reply:
I don't think a key can be null here since it's really just a Map under the hood

Reply:
Java's HashMap<String, String> allows a key to be null.

Reply:
I pulled these checks from the `newHadoopConfWithOptions ` method (which takes in a scala.Map). It checks for `v != null && k != "path" && k != "paths"`. So I simplified it to `v != null and k.startsWith`.

The reason I didn't use that function is to avoid having to convert back and forth between java Map and scala Map (which was quite ugly comparatively speaking).

Reply:
I can leave a comment there if we'd like about where I pulled that from?

Reply:
I left a comment re: where these checks came from.

It's possible that maybe we _do_ want to allow for `null` values so users can unset configs that are set on the default hadoop configuration for the session. But I'm not sure if setting to `null` would correctly revert it to the default value. I think they'd need to explicitly set it to whatever default value they want.

Reply:
I can check `k` for null, but it wasn't checked in the `newHadoopConfWithOptions`. Possibly because `scala.Map` cannot have `null` keys?

Reply:
I was able to put a `null` key into a `scala.Map[String, String]`.
```
scala> var nullString: String = null
nullString: String = null

scala> x += nullString -> "5"
res3: scala.collection.mutable.Map[String,String] = Map(null -> 5)

scala> x
res4: scala.collection.mutable.Map[String,String] = Map(null -> 5)
```

However, putting a `null` key into the hadoop configuration throws:
```
scala> var config = spark.sessionState.newHadoopConf
config: org.apache.hadoop.conf.Configuration = Configuration: core-default.xml, core-site.xml, mapred-default.xml, mapred-site.xml, yarn-default.xml, yarn-site.xml, hdfs-default.xml, hdfs-site.xml, __spark_hadoop_conf__.xml

scala> config.set(null, "10")
java.lang.IllegalArgumentException: Property name must not be null
  at com.google.common.base.Preconditions.checkArgument(Preconditions.java:92)
  at org.apache.hadoop.conf.Configuration.set(Configuration.java:1353)
  at org.apache.hadoop.conf.Configuration.set(Configuration.java:1337)
  ... 47 elided
```

I think that `settings` shouldn't return a `null` key, but I can add a check just in case if we think it's a good idea.

Reply:
Thanks for the test. For checking if k is null, it's OK to leave it, but safer to add it.

Reply:
Ok. Happy to add it. I can't imagine there's any practical performance reason not to, and this function will only be invoked a few times anyway.

Coming from Scala, I'm always scared of null anyways 😜

Reply:
Added a check for `k != null` as well. 🙂